presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBatch.java

```diff
@@ -242,4 +384,62 @@ private RowTupleSupplier(int partitionCount, int rowCount, byte[] rowData, int[]
             return tuple;
         }
     }
+
+    public static class RowIndex
```

Comment:
The main reason for implementing RowIndex was to avoid allocating extra objects along the way.

Let me add a comment explaining how it works at the top of the class.